File: brain-js/workspaces/codex/examples/investigate.ts
import { z } from "zod";

import { Internet, complete, fail, flow, next, node } from "brain-js";
import { CodexRuntime } from "brain-js-codex";

interface ResearchInput {
  topic: string;
}

const researchResult = z.object({ finding: z.string() });

interface AnalysisInput {
  finding: string;
}

const analysisResult = z.object({
  report: z.string(),
  followUp: z.string().nullable(),
});

const topic = process.argv[2] ?? "typed agent workflows";
const research = node<ResearchInput, z.infer<typeof researchResult>>(
  "research",
  researchResult,
)
  .prompt("Research the topic and return one important, well-supported finding.")
  .internet(Internet.Enabled);
const analyze = node<AnalysisInput, z.infer<typeof analysisResult>>(
  "analyze",
  analysisResult,
).prompt(
  "Analyze the finding. Return a final report, or a focused follow-up topic if more research is needed.",
);
const analyzeAfterResearch = analyze.clone();
const researchAfterAnalysis = research.clone();

const run = await flow<string>("investigate")
  .beginsWith(research.with({ topic }))
  .after(research, (outcome) =>
    outcome.ok
      ? next(analyzeAfterResearch.with({ finding: outcome.value.finding }))
      : fail(outcome.failure.intoError()),
  )
  .after(analyze, (outcome) => {
    if (!outcome.ok) return fail(outcome.failure.intoError());
    return outcome.value.followUp === null
      ? complete(outcome.value.report)
      : next(researchAfterAnalysis.with({ topic: outcome.value.followUp }));
  })
  .run(new CodexRuntime());

console.log(run.output);

File: brain/crates/codex/examples/investigate.rs
use brain::{FlowError, Internet, complete, fail, flow, next, node};
use brain_codex::CodexRuntime;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Serialize)]
struct ResearchInput {
    topic: String,
}

#[derive(Deserialize, JsonSchema)]
struct ResearchResult {
    finding: String,
}

#[derive(Serialize)]
struct AnalysisInput {
    finding: String,
}

#[derive(Deserialize, JsonSchema)]
struct AnalysisResult {
    report: String,
    follow_up: Option<String>,
}

#[tokio::main]
async fn main() -> Result<(), FlowError> {
    let topic = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "typed agent workflows".into());

    let research = node::<ResearchInput, ResearchResult>("research")
        .prompt("Research the topic and return one important, well-supported finding.")
        .internet(Internet::Enabled);
    let analyze = node::<AnalysisInput, AnalysisResult>("analyze")
        .prompt("Analyze the finding. Return a final report, or a focused follow-up topic if more research is needed.");

    let run = flow::<String>("investigate")
        .begins_with(research.with(ResearchInput { topic }))
        .after(research, move |outcome| match outcome {
            Ok(result) => next(analyze.with(AnalysisInput {
                finding: result.finding,
            })),
            Err(failure) => fail(failure.into_error()),
        })
        .after(analyze, move |outcome| match outcome {
            Ok(result) => match result.follow_up {
                Some(topic) => next(research.with(ResearchInput { topic })),
                None => complete(result.report),
            },
            Err(failure) => fail(failure.into_error()),
        })
        .run(&CodexRuntime::new())
        .await?;

    println!("{}", run.output);
    Ok(())
}
